wgpu-hal/src/gles/device.rs

```diff
@@ -1133,23 +1180,40 @@ impl crate::Device<super::Api> for super::Device {
         })
     }
     unsafe fn destroy_render_pipeline(&self, pipeline: super::RenderPipeline) {
-        let gl = &self.shared.context.lock();
-        unsafe { gl.delete_program(pipeline.inner.program) };
+        let mut program_cache = self.shared.program_cache.lock();
+        // 1 means we're the last thing holding this, so we can safely clean it up
+        if Arc::strong_count(&pipeline.inner) == 1 {
```

Comment:
This is probably correct due to the program cache lock, but if that wasn't there, this would be ever so subtly wrong. Arc::try_unwrap would get rid of that issue by doing the decrement and the extraction in the same step. It's a bit messier syntactically, but much more clearly correct.

Reply:
If the lock weren't there it wouldn't compile 😅
I can try that but you're right that it's going to be a little messy.

The actual last holder is inside the cache, but we'll need to drop the parameter `pipeline` to ensure it'll be the final final one. `Arc::try_unwrap` requires passing in an owned arc, not a ref, so we'd need to take it out of the cache but we can't/shouldn't take it out of the cache until we know it's the last one.

(I just noticed with the change to Arc refcount instead of the owned refcount, behaviour slightly changed: now it's counting the users handle of a `RenderPipeline`, whereas before I didn't care about that. That means the `1` should actually be `2` here, or I explicitly drop the passed-in pipeline first.)

I can only think of taking out of the cache, doing the test, and adding it back in - but that's even more error prone. What do you think?

Reply:
Alright, yeah that seems quite a bit more complicated 😅. Sticking with strong count is fine then, but could you add comments to this line (and things that increase the refcount) why it's correct?

Reply:
I believe one of you needs to mark this conversation as resolved for the auto-merge to go through.

Reply:
Thanks! Didn't realize it wasn't resolved